runtime/src/tests/staking.rs

```diff
@@ -72,3 +77,463 @@ fn stake_and_unstake_with_a_controller_and_stash_as_payee() {
+		));
+
+		assert_noop!(
+			Staking::set_controller(
+				RuntimeOrigin::signed(stash),
+				sp_runtime::MultiAddress::Id(controller)
+			),
+			Error::<Runtime>::AlreadyPaired
+		);
+	});
+}
+
+#[test]
+fn set_controller_should_work() {
+	let stash = generate_account("stash");
+	let controller = generate_account("controller");
+	let new_controller = generate_account("new-controller");
+	let value = <Balances as Currency<AccountId>>::minimum_balance();
+
+	ExtBuilder::<()>::default().build_sans_config().execute_with(|| {
+		let mut height = 1;
+		roll_to::<Trivial, Runtime, ()>(height);
+		Timestamp::set_timestamp(1);
+		//bypass defensive first era.
+		{
+			staking::ActiveEra::<Runtime>::set(Some(ActiveEraInfo { index: 0, start: Some(0) }));
+			staking::ErasStartSessionIndex::<Runtime>::set(0, Some(height));
+		}
+
+		height += BlocksPerEra::get();
+		assert_eq!(height, Staking::next_era_start(1));
+		roll_to::<Trivial, Runtime, TaskVoting>(height);
+
+		//prime balance with enough funds to prevent dusting
+		let _ = Balances::deposit_creating(&stash, 2 * value);
+
+		assert_ok!(Staking::bond(
+			RuntimeOrigin::signed(stash.clone()),
+			controller.clone().into(),
+			value,
+			RewardDestination::Stash,
+		));
+		assert_eq!(controller, Staking::bonded(stash.clone()).unwrap());
+		assert_matches!(Staking::ledger(&controller), Some(_));
+		assert_matches!(Staking::ledger(&new_controller), None);
+
+		assert_ok!(Staking::set_controller(
+			RuntimeOrigin::signed(stash.clone()),
+			sp_runtime::MultiAddress::Id(new_controller.clone())
+		));
+
+		assert_eq!(new_controller, Staking::bonded(stash).unwrap());
+		assert_matches!(Staking::ledger(&controller), None);
+		assert_matches!(Staking::ledger(&new_controller), Some(_));
+	});
+}
+
+#[test]
+fn set_invulnerables_should_be_signed() {
+	ExtBuilder::<()>::default().build_sans_config().execute_with(|| {
+		assert_noop!(Staking::set_invulnerables(RuntimeOrigin::none(), vec![]), BadOrigin);
+	})
+}
+
+#[test]
+fn set_invulnerables_should_be_signed_by_root() {
+	let controller = generate_account("not-root");
+
+	ExtBuilder::<()>::default().build_sans_config().execute_with(|| {
+		assert_noop!(
+			Staking::set_invulnerables(RuntimeOrigin::signed(controller), vec![]),
+			BadOrigin
+		);
+	})
+}
+
+#[test]
+fn set_invulnerables_should_work() {
+	let alex = generate_account("Alex");
+	let bob = generate_account("Bob");
+
+	ExtBuilder::<()>::default().build_sans_config().execute_with(|| {
+		assert_eq!(Staking::invulnerables(), vec![]);
+
+		assert_ok!(Staking::set_invulnerables(
+			RuntimeOrigin::root(),
+			vec![alex.clone(), bob.clone()]
+		));
+		assert_eq!(Staking::invulnerables(), vec![alex.clone(), bob.clone()]);
+
+		// making another call should update the list of invulnerable accounts
+		let charlie = generate_account("charlie");
+		assert_ok!(Staking::set_invulnerables(RuntimeOrigin::root(), vec![charlie.clone()]));
+		assert_eq!(Staking::invulnerables(), vec![alex, bob, charlie]);
```

Comment:
Currently this assertion fails:

```
failures:

---- tests::staking::set_invulnerables_should_work stdout ----
thread 'tests::staking::set_invulnerables_should_work' panicked at 'assertion failed: `(left == right)`
  left: `[a072b7a5c262bab8b34511eebdf0d00cc03202403394593c2e243065e6c3f976 (5Fh5baCG...)]`,
 right: `[dbc63fb2926c6e95dbb24964acdfd17391c513a5d47e2f44afaf1f5861751a22 (5H2sEeXk...), 81855c93f0d51e369107404b7a1f76795b1d9634384d926dbb6260eac958233d (5EzXei7A...), a072b7a5c262bab8b34511eebdf0d00cc03202403394593c2e243065e6c3f976 (5Fh5baCG...)]`', runtime/src/tests/staking.rs:537:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

@CAGS295 do we want to overwrite the list of invulnerable accounts when `set_invulnerables()` is executed or do we want to append to it?

Reply:
Inconclusive yet, this logic depends on how we want to transition away from the federated consensus..

Reply:
I think the simplest approach for now is to overwrite.